Experimental class bindings

AngelMunoz opened the pull request:
This is the class bindings PR

these are the first bits but I want to iterate over this and see if there's a way we can also apply some of these concepts to the LitElement decorated functions

I'm still missing some bindings like Directives and AsyncDirectives (they just have what they need now, but ideally they should be fully typed)

BillHally committed:
Fix a typo

AngelMunoz committed:
add styles in the Lit functions

alfonsogarciacaro committed:
Remove mention to --runFast in docs
(Not necessary since Fable 3.7)

alfonsogarciacaro committed:
Solutions for #31 and #39

AngelMunoz committed:
add extra types to LitElement

AngelMunoz committed:
add computation expressions for Class LitElements

AngelMunoz committed:
add class based examples

AngelMunoz committed:
update to .NET6

AngelMunoz committed:
update local tools
add fantomas

AngelMunoz committed:
re-order things a bit

AngelMunoz committed:
fix format and simplify CEs

alfonsogarciacaro commented on sample/App.fs:
```diff
@@ -49,10 +61,30 @@ let view model dispatch =
+    view model dispatch
+
+[<Emit("String")>]
+let StringCtor: obj = jsNative
```
There are helpers for the prop types here: https://github.com/fable-compiler/Fable.Lit/blob/6317d3a32ed91927984b1fc03aa7da3fc02affa3/src/Lit/LitElement.fs#L24-L30

But if they are going to be public I would use an enumeration and then translate it internally or at least fake so users don't pass any kind of object:

```fsharp
type PropType =
    [<Global>] static member String: PropType = jsNative
    [<Global>] static member Number: PropType = jsNative
    [<Global>] static member Boolean: PropType = jsNative
    [<Global>] static member Array: PropType = jsNative
    [<Global>] static member Object: PropType = jsNative
```

alfonsogarciacaro commented on sample/App.fs:
```diff
@@ -49,10 +61,30 @@ let view model dispatch =
+[<Emit("String")>]
+let StringCtor: obj = jsNative
+
+registerElement "user-profile" JsInterop.jsConstructor<ClassComponents.UserProfile> {
```
Not sure how it works with CE but I would provide an inline generic version so users don't need to call `jsConstructor` like:

```fsharp
// This is equivalent to the current registerElement, cannot be private because the new registerElement is inlined 
// but shouldn't be used directly
let registerElementFromConstructor (name: string) (cons: obj) = ...

let inline registerElement<'T> (name: string) = registerElementFromConstructor name jsConstructor<'T>
```

AngelMunoz replied:
that feels better, yes we can do that

alfonsogarciacaro commented on sample/App.fs:
```diff
@@ -49,10 +61,30 @@ let view model dispatch =
+
+registerElement "element-with-controller" JsInterop.jsConstructor<ClassComponents.ElementWithController> {
+    css $"p {{ color: red; }}"
+    css $"li {{ color: rebeccapurple; }}"
```
A bit concerned that the class and the registration (with the props, styles...) are so far away so the compiler cannot tell you when you misspell a prop name. But let's see how it goes (it's that compiler doesn't the prop names of html templates either).

AngelMunoz replied:
I don't think this is too far fetched from what current lit users do, but I share this concern arguably I think it would be nice if we stick slightly more close to lit's semantics so people don't have to learn Fable.Lit + Lit

alfonsogarciacaro commented on sample/ClassComponents.fs:
```diff
@@ -0,0 +1,74 @@
+type Post = {| usedId: int; id: int; title: string; body: string |}
+
+[<Import("MouseController", from="./controllers.js"); Emit("new $0($1)")>]
+let private createMouseController (host: LitElement): MouseController = jsNative
```
If you need the constructor, instead of using an interface and a separate function, it's probably easier to declare a class:

```fsharp
[<ImportMember("./controllers.js")>]
type MouseController(host: ReactiveControllerHost) =
    inherit ReactiveController(host)
    member _.x: float = jsNative
    member _.y: float = jsNative
```

AngelMunoz replied:
AH I just copied this from the controller sample we had already on the branch I will remove it before I finish this

alfonsogarciacaro commented on sample/ClassComponents.fs:
```diff
@@ -0,0 +1,74 @@
+
+type ApiController(host: LitElement) as this =
+    do
+        host.addController(this)
```
I don't like to use `as this` very much in constructors, because it often raises warnings from the F# compiler and the code generated is quite ugly. If we declare `ReactiveController` as an abstract class we can do this automatically and also add default implementations for the abstract members so users don't need to implement all of them:

```fsharp
[<AbstractClass; AttachMembers>]
type ReactiveController(host: ReactiveControllerHost) =
    do host.addController(jsThis)
    member _.host = host

    abstract hostConnected: unit -> unit
    default _.hostConnected() = ()

    abstract hostDisconnected: unit -> unit
    default _.hostDisconnected() = ()

    abstract hostUpdate: unit -> unit
    default _.hostUpdate() = ()

    abstract hostUpdated: unit -> unit
    default _.hostUpdated() = ()
```

Then your sample becomes:

```fsharp
type ApiController(host: ReactiveControllerHost) =
    inherit ReactiveController(host)
    override this.hostConnected() = printfn "Connected!"
```

AngelMunoz committed:
fix sample usage

AngelMunoz committed:
inline CE, address pr comments
expose PropTypes

AngelMunoz committed:
fix reactive controller call

AngelMunoz committed:
update definitions
ReactiveController, ReactiveControllerHost

AngelMunoz committed:
update lit version

AngelMunoz committed:
fix pr comments

AngelMunoz committed:
add full bindings to directives

this should give anyone the ability to create full class directives if they need it

AngelMunoz committed:
add experimental module

AngelMunoz commented on sample/App.fs:
```diff
@@ -49,10 +63,58 @@ let view model dispatch =
+
+[<LitElementExperimental(true)>]
+let Sample () =
+    let name = Controllers.GetProperty("name", "Peter")
+
+    let state: StateController<int> =
+        Controllers.GetController<StateController<int>>(10)
+
+    let twoArgs = Controllers.GetController<TwoArgsCtrl>(10, 20)
```
I was thinking on something like this to prevent having the LitElement generic initialization, we still do magic, but more _javascrip'ty_ magic (or that's what I want to believe) in the end these are just properties that are in the class that is being used behind this function to use lit's mechanisms without diving too much into directive territory

Side note perhaps _**Context**_ would be a better name rather than _Controllers_

AngelMunoz commented on src/Lit/Experimental.fs:
```diff
@@ -0,0 +1,76 @@
+[<Experimental("This is an exploration space, anything can be added or removed in this module at any point")>]
+module Lit.Experimental
```
I want to learn how you did actually do the Decoration stuff as well as hot reload so any experiments and ideas I have in mind I will put them right here, hope you don't mind and also if you see things we can improve and take out to the real world stuff it would be nice

AngelMunoz commented on src/Lit/Experimental.fs:
```diff
@@ -0,0 +1,76 @@
+open Lit
+
+[<AttachMembers>]
+type StateController<'T>(host, initial: 'T) =
+    inherit ReactiveController(host)
+    let mutable state = initial
+
+    member _.Value = state
+
+    member _.SetState(value: 'T) =
+        state <- value
+        host.requestUpdate ()
```
I'm playing with the idea of having hook-like abstractions but rather than backed by a directive backed by ReactiveControllers they would work nicely for both class function based LitElements, too crazy perhaps?

AngelMunoz commented on src/Lit/Experimental.fs:
```diff
@@ -0,0 +1,76 @@
+        host.requestUpdate ()
+
+[<AttachMembers>]
+type TwoArgsCtrl(host, initial: int, secondial: int) =
```
don't mind the bad english names, I was just testing `System.ParamArray` and apply function arguments to the constructor

AngelMunoz commented on src/Lit/Experimental.fs:
```diff
@@ -0,0 +1,76 @@
+        host.requestUpdate ()
+
+
+type Controllers =
+
+    static member inline GetController<'T>([<System.ParamArray>] args: obj array) : 'T =
+        jsThis?getOrAddController (jsConstructor<'T>, args)
+
+    static member inline GetProperty(propName: string, ?initial: 'T) : Types.RefValue<'T> =
+        jsThis?getOrAddProperty (propName, initial)
```
As I was saying in a comment above, Context would be a better name for this rather than controllers

This plays on the idea that Function Based LitElements have a class that can store either properties/controllers or similar things so rather than have an initialize function we should be able to grab them from the backing class of those decorated functions.

This is kind of crafted for my own experiment but I think we could somehow make it work with your existing LitElement functions without taking away the existing support for the init function this no breaking changes + a way to progress forward

AngelMunoz commented on src/Lit/Experimental.fs:
```diff
@@ -0,0 +1,76 @@
+        jsThis?getOrAddProperty (propName, initial)
+
+[<Literal>]
+let CLS_EXPR =
+    """class extends $0 {
+        constructor() {
+            super();
+            this.renderFn = $1
+        }
+
+        render() {
+            return this.renderFn.value.apply(this, [])
+        }
+
+        getOrAddController(controller, args) {
+            if(!this[controller.name]) {
+                return this[controller.name] = new controller(this, ...args)
+            }
+            return this[controller.name];
+        }
+
+        getOrAddProperty(propname, initial) {
+            if(!this[propname]) {
+                this[propname] = initial;
+                return this[propname];
+            }
+            return this[propname];
+        }
+    }
+    """
+
+type LitElementExperimental(useShadowDom: bool) =
+    inherit JS.DecoratorAttribute()
+
+    override this.Decorate(fn: JS.Function) : JS.Function =
+        let renderRef = LitBindings.createRef ()
+        renderRef.value <- fn
+        let baseClass = emitJsExpr (jsConstructor<LitElement>, renderRef) (CLS_EXPR)
+
+        baseClass
```
This is me just trying to learn how to use the decorator function so I don't break your existing stuff while I learn how it works and experiment a bit with it so I can get familiar with the code base

I think (not entirely sure yet) that we could even decorate classes this way and also allow for hot reload stuff but I'm still away from it

AngelMunoz reviewed (commented):
@alfonsogarciacaro I know you have a quite busy agenda but if any of this makes sense or not feel free to let me know when you have time, I really want to get a deeper understanding of this codebase so I can keep taking it further away

AngelMunoz committed:
remove decorator experiment
add effect controller

AngelMunoz committed:
add elmish controller test

AngelMunoz committed:
enable controllers + hook.GetController

leolorenzoluis committed:
Fixing self closing web component tags

leolorenzoluis committed:
Added Lit.mapiUnique

alfonsogarciacaro committed:
Publish 1.4.2

alfonsogarciacaro committed:
Update templates.md

AngelMunoz committed:
Merge branch 'master' into experimental-class-bindings

AngelMunoz committed:
Merge branch 'master' into experimental-class-bindings

JordanMarr committed:
Update hook-components.md

Fixed error in custom hook example.

alfonsogarciacaro committed:
Update hook-components.md

alfonsogarciacaro committed:
Fix #67

alfonsogarciacaro committed:
Fix Github actions

AngelMunoz committed:
Merge branch 'master' into experimental-class-bindings

AngelMunoz committed:
change the name for consistency

AngelMunoz commented:
Hey @alfonsogarciacaro I don't really have more to add to this PR I think it has been idle for a while but the important bits are there.

I added the `useController` so users that are using function based lit elements can also grab controllers that may already exist in the wild but ideally the controllers match better with the class based API which mirrors what actual lit users do and might be simpler for some to copy or port existing components
I will be adding docs for these today and then feel free to review

AngelMunoz committed:
add class docs

AngelMunoz committed:
add controllers docs

AngelMunoz committed:
add new pages to the menu

AngelMunoz committed:
remove the effect controller

AngelMunoz committed:
fix the compilation issue

AngelMunoz committed:
add doc strings

AngelMunoz committed:
add a couple more comments

goswinr commented:
@alfonsogarciacaro thanks you for making Fable .Lit 🙏 I am just learning Lit and wondered why I can't do this example in Fable.Lit: https://lit.dev/docs/composition/controllers/#example:-mousemovecontroller
I found the answer here in this pull request. Thanks, @AngelMunoz.  Any reason not to merge this ? At least to have the official samples working in Fable.Lit too?

AngelMunoz commented:
No particular reason I never got to finish this as I got distracted.

I'd like more feedback on this as well if you want to try these bits we could revisit the work and make the appropriate updates if needed 

Cc: @JordanMarr , @OnurGumus

Sorry to bother I've seen you are using this as on Twitter as well I guess this could be of interest

OnurGumus commented:
I think the current approach is fine. One can just use the controllers the way you do. So far I haven't seen any need for this PR.

JordanMarr commented:
Sorry, all my free coding time has been consumed with the Elmish.Avalonia project (and my Fable.Lit project got iced). 
TBH, I don't even know what this PR is about as there are no pretty pictures of code look at. 
Based on Onur's comment, I'm assuming that you created bindings for Lit [Controllers](https://lit.dev/docs/composition/controllers/)?
